test/framework/ssh/portforward.go

```diff
@@ -0,0 +1,58 @@
+package ssh
+
+import (
+	"io"
+	"log"
+	"net"
+
+	"golang.org/x/crypto/ssh"
+)
+
+func Tunnel(sshClient *ssh.Client, local, remote string, stopChan <-chan struct{}, readyChan chan<- net.Addr) error {
+	pipe := func(writer, reader net.Conn) {
+		defer func(writer net.Conn) {
+			_ = writer.Close()
+		}(writer)
+		defer func(reader net.Conn) {
+			_ = reader.Close()
+		}(reader)
+
+		_, err := io.Copy(writer, reader)
+		if err != nil {
+			log.Printf("port forward failed: %s", err)
+		}
+	}
+
+	listener, err := net.Listen("tcp", local)
+	if err != nil {
+		return err
+	}
+
+	if readyChan != nil {
+		readyChan <- listener.Addr()
+	}
+
+	for {
+		select {
+		case <-stopChan:
+			_ = listener.Close()
+			return nil
+		default:
+			localConnection, err := listener.Accept()
+			if err != nil {
+				return err
+			}
+
+			go func(localConn net.Conn) {
+				remoteConn, err := sshClient.Dial("tcp", remote)
+				if err != nil {
+					log.Fatalf("failed to dial to ssh remote %q on host %s: %q",
+						remote, sshClient.RemoteAddr().String(), err)
+				}
+
+				go pipe(remoteConn, localConn)
+				go pipe(localConn, remoteConn)
```

Comment:
I don't understand this. Doesn't this create some sort of a loop? 🤔

Reply:
Sockets are bi-directional. So a socket connection can be used as both `io.Reader` and `io.Writer`.

The `pipe` function takes two parameter as `io.Writer` and `io.Reader`. So a connection is first used as an `io.Writer` and then used as an `io.Reader`, or the other way around.

Each invocation handles a single direction, so there is no loop.